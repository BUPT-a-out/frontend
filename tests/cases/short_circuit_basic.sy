int main() {
    int a = 0;
    int b = 1;
    int result = 0;
    
    if (a && b) {
        result = 1;
    }
    
    if (b || a) {
        result = result + 10;
    }
    
    return result;
}