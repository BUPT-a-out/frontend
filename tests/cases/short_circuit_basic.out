Parsing completed successfully.

--- Abstract Syntax Tree ---
+-- ROOT
|   +-- FUNC_DEF: main (sym: main, id: 26)
|   |   +-- LIST: FParams
|   |   +-- LIST: Block
|   |   |   +-- VAR_DEF: a (sym: a, id: 27)
|   |   |   |   +-- CONST (int value: 0)
|   |   |   +-- VAR_DEF: b (sym: b, id: 28)
|   |   |   |   +-- CONST (int value: 1)
|   |   |   +-- VAR_DEF: result (sym: result, id: 29)
|   |   |   |   +-- CONST (int value: 0)
|   |   |   +-- IF_STMT
|   |   |   |   +-- BINARY_OP: &&
|   |   |   |   |   +-- VAR (sym: a, id: 27)
|   |   |   |   |   +-- VAR (sym: b, id: 28)
|   |   |   |   +-- LIST: If-2
|   |   |   |   |   +-- ASSIGN_STMT
|   |   |   |   |   |   +-- VAR (sym: result, id: 29)
|   |   |   |   |   |   +-- CONST (int value: 1)
|   |   |   +-- IF_STMT
|   |   |   |   +-- BINARY_OP: ||
|   |   |   |   |   +-- VAR (sym: b, id: 28)
|   |   |   |   |   +-- VAR (sym: a, id: 27)
|   |   |   |   +-- LIST: If-2
|   |   |   |   |   +-- ASSIGN_STMT
|   |   |   |   |   |   +-- VAR (sym: result, id: 29)
|   |   |   |   |   |   +-- BINARY_OP: +
|   |   |   |   |   |   |   +-- VAR (sym: result, id: 29)
|   |   |   |   |   |   |   +-- CONST (int value: 10)
|   |   |   +-- IF_STMT
|   |   |   |   +-- UNARY_OP: !
|   |   |   |   |   +-- VAR (sym: result, id: 29)
|   |   |   |   +-- LIST: If-2
|   |   |   |   |   +-- RETURN_STMT: main (sym: main, id: 26)
|   |   |   |   |   |   +-- VAR (sym: b, id: 28)
|   |   |   +-- RETURN_STMT: main (sym: main, id: 26)
|   |   |   |   +-- VAR (sym: result, id: 29)

--- Permanent Symbol Table ---
ID    Name                 Type            Data Type  Function             Shape     
-------------------------------------------------------------------------------------
0     getint               function        int        N/A                  N/A       
1     getch                function        int        N/A                  N/A       
2     getfloat             function        float      N/A                  N/A       
3     getarray             function        int        N/A                  N/A       
4     array                array           int        getarray             0
5     getfarray            function        int        N/A                  N/A       
6     array                array           float      getfarray            0
7     putint               function        void       N/A                  N/A       
8     value                var             int        putint               N/A       
9     putch                function        void       N/A                  N/A       
10    value                var             int        putch                N/A       
11    putfloat             function        void       N/A                  N/A       
12    value                var             float      putfloat             N/A       
13    putarray             function        void       N/A                  N/A       
14    len                  var             int        putarray             N/A       
15    array                array           int        putarray             0
16    putfarray            function        void       N/A                  N/A       
17    len                  var             int        putfarray            N/A       
18    array                array           float      putfarray            0
19    putf                 function        void       N/A                  N/A       
20    format_string        var             char       putf                 N/A       
21    value                var             int        putf                 N/A       
22    starttime            function        void       N/A                  N/A       
23    line                 var             int        starttime            N/A       
24    stoptime             function        void       N/A                  N/A       
25    line                 var             int        stoptime             N/A       
26    main                 function        int        N/A                  N/A       
27    a                    var             int        main                 N/A       
28    b                    var             int        main                 N/A       
29    result               var             int        main                 N/A       
-------------------------------------------------------------------------------------

--- Generated IR ---
; ModuleID = 'main'

define i32 @main() {
main.entry:
  %a.27 = alloca i32
  %b.28 = alloca i32
  %result.29 = alloca i32
  store i32 0, i32* %a.27
  store i32 1, i32* %b.28
  store i32 0, i32* %result.29
  %0 = load i32, i32* %a.27
  br i1 %0, label %and.1.rhs, label %and.1.merge
and.1.rhs:
  %1 = load i32, i32* %b.28
  br label %and.1.merge
and.1.merge:
  %and.1.result = phi i1 [ 0, %main.entry ], [ %1, %and.1.rhs ]
  br i1 %and.1.result, label %if.0.then, label %if.0.merge
if.0.then:
  store i32 1, i32* %result.29
  br label %if.0.merge
if.0.merge:
  %2 = load i32, i32* %b.28
  br i1 %2, label %or.3.merge, label %or.3.rhs
or.3.rhs:
  %3 = load i32, i32* %a.27
  br label %or.3.merge
or.3.merge:
  %or.3.result = phi i1 [ 1, %if.0.merge ], [ %3, %or.3.rhs ]
  br i1 %or.3.result, label %if.2.then, label %if.2.merge
if.2.then:
  %4 = load i32, i32* %result.29
  %add.5 = add i32 %4, 10
  store i32 %add.5, i32* %result.29
  br label %if.2.merge
if.2.merge:
  %6 = load i32, i32* %result.29
  %not.7 = icmp eq i32 %6, 0
  br i1 %not.7, label %if.4.then, label %if.4.merge
if.4.then:
  %8 = load i32, i32* %b.28
  ret i32 %8
if.4.merge:
  %9 = load i32, i32* %result.29
  ret i32 %9
}


